sycl/include/sycl/detail/is_device_copyable.hpp

```diff
@@ -100,14 +100,24 @@ struct __SYCL2020_DEPRECATED("This type isn't device copyable in SYCL 2020")
     IsDeprecatedDeviceCopyable<T[N]> : IsDeprecatedDeviceCopyable<T> {};
 
 #ifdef __SYCL_DEVICE_ONLY__
+// Helper trait to avoid instantiated the deprecated specializations of
+// IsDeprecatedDeviceCopyable when is_device_copyable_v<T> is true.
+template <typename T, typename = void>
+struct IsDeprecatedDeviceCopyableIfNot2020 : std::false_type {};
+
+template <typename T>
+struct IsDeprecatedDeviceCopyableIfNot2020<
+    T, std::enable_if_t<!is_device_copyable_v<T>>>
+    : detail::IsDeprecatedDeviceCopyable<T> {};
+
 // Checks that the fields of the type T with indices 0 to (NumFieldsToCheck -
 // 1) are device copyable.
 template <typename T, unsigned NumFieldsToCheck>
 struct CheckFieldsAreDeviceCopyable
     : CheckFieldsAreDeviceCopyable<T, NumFieldsToCheck - 1> {
   using FieldT = decltype(__builtin_field_type(T, NumFieldsToCheck - 1));
   static_assert(is_device_copyable_v<FieldT> ||
-                    detail::IsDeprecatedDeviceCopyable<FieldT>::value,
+                    detail::IsDeprecatedDeviceCopyableIfNot2020<FieldT>::value,
```

Comment:
Any updates on this?

Reply:
Personally I would prefer that we drop it in a separate patch, just to make it easier to bisect in case it breaks something unintentionally.

Reply:
> Personally I would prefer that we drop it in a separate patch, just to make it easier to bisect in case it breaks something unintentionally.

I have no objections against doing that in a separate PR and I agree that it may be easier for us in a potential future bisect when some customer complains that their code doesn't compile/work anymore